### PR TITLE
Avoid recursive algorithm, expand complex rib test, and include and/or

### LINKED ISSUE #986: Expand the comprehensive test to cover all scenarios
It looks like with the new [mock_data](https://github.com/golemcloud/golem/blob/migrate_tests/golem-rib/src/interpreter/tests/mod.rs#L1191), [mock_interpreter](https://github.com/golemcloud/golem/blob/migrate_tests/golem-rib/src/interpreter/tests/mod.rs#L1440), [data-types](https://github.com/golemcloud/golem/blob/migrate_tests/golem-rib/src/interpreter/tests/mod.rs#L754) [extensive function metadata](https://github.com/golemcloud/golem/blob/migrate_tests/golem-rib/src/interpreter/tests/mod.rs#L432) and [component metadata](https://github.com/golemcloud/golem/blob/migrate_tests/golem-rib/src/interpreter/tests/mod.rs#L385), its now easier to expand on [test_interpreter_complex_rib](https://github.com/golemcloud/golem/blob/migrate_tests/golem-rib/src/interpreter/tests/mod.rs#L8) which is a dedicated module to test a complex Rib. 

While we have dedicated test for each functionality, its important to test how it behaves when chaining them altogether, such as taking the result of a function and construct a variant.  

Currently it looks like only 70%, which is already significantly better than what existed during the initial stages of Rib.

### PR DESCRIPTION
**Pull request by afsalthaj**
Lot many subtle bugs captured in this PR, and I will try to explain them in PR description if possible

Fixes #986  #782 and #893

### CODE REVIEW

File: golem-rib/src/interpreter/tests/mod.rs [hunk 1]
```diff
@@ -169,17 +169,17 @@ mod comprehensive_test {
...
                 ok(number) => number,
-                err(msg) => 0
+                err(number) => number
               };
 
               let result_of_variant_response_processed = match result_of_variant_response {
-                ok(case-str(_)) => "found",
-                err(msg) => "not found"
+                ok(case-str(a)) => a,
+                err(case-str(b)) => b
```
Comment (afsalthaj) [inline golem-rib/src/interpreter/tests/mod.rs:182]: These never used to work!

File: golem-rib/src/interpreter/tests/mod.rs [hunk 2]
```diff
@@ -262,11 +262,15 @@ mod comprehensive_test {
         let mut rib_executor = mock_interpreter::interpreter();
         let result = rib_executor.run(compiled_expr).await.unwrap();
 
+        dbg!(test_utils::convert_type_annotated_value_to_str(
+            &result.get_val().unwrap()
+        ));
+
```
Comment (afsalthaj) [inline golem-rib/src/interpreter/tests/mod.rs:268]: Add assertion based on wasm wave str, to make it debugging easier in future.

File: golem-rib/src/type_inference/type_pull_up.rs [hunk 3]
```diff
@@ -13,187 +13,697 @@
...
-//   * Fill up a queue with the root node being first
-//  [select_field(select_field(a, b), c), select_field(a, b), identifier(a)]
-//  Pop from back and push to the front of a stack of the current expression's inferred type, and keep assigning in between
-// Example:
-//  * Pop back to get identifier(a)
-//  * Try to pop_front inferred_type_stack, and its None. Push front the identifier(a)'s inferred_type:  Record(b -> Record(c -> u64))
-//  * Pop back from stack to get select_field(a, b)
-//  * Try to pop_front inferred_type_stack, and its Record(b -> Record(c -> u64)). Get the type of b and assign itself and push_front to stack.
-//  * Pop back from stack to get select_field(select_field(a, b), c)
-//  * Try to pop_front inferred_type_stack, and its  Record(c -> u64). Get the type of c and assign itself and push to stack.
```
Comment (afsalthaj) [inline golem-rib/src/type_inference/type_pull_up.rs:29]: This is the logic that's implemented in this PR

**Comment by afsalthaj** [comment]
Fix #986  #782  and #893

**Review by jdegoes — APPROVED** [review]

**Pull request merged**